perception/autoware_ground_segmentation/src/scan_ground_filter/grid_ground_filter.cpp

```diff
@@ -0,0 +1,460 @@
+    idx.push_back(check_idx);
+    return recursiveSearch(check_cell.scan_grid_root_idx_, search_cnt - 1, idx);
+  }
+  // if the cell does not have ground, search previous cell
+  return recursiveSearch(check_cell.scan_grid_root_idx_, search_cnt, idx);
+}
+
+void GridGroundFilter::fitLineFromGndGrid(const std::vector<int> & idx, float & a, float & b) const
+{
+  // if the idx is empty, the line is not defined
+  if (idx.empty()) {
+    a = 0.0f;
+    b = 0.0f;
+    return;
+  }
+  // if the idx is length of 1, the line is zero-crossing line
+  if (idx.size() == 1) {
+    const auto & cell = grid_ptr_->getCell(idx.front());
+    a = cell.avg_height_ / cell.avg_radius_;
+    b = 0.0f;
+    return;
+  }
+  // calculate the line by least square method
+  float sum_x = 0.0f;
+  float sum_y = 0.0f;
+  float sum_xy = 0.0f;
+  float sum_x2 = 0.0f;
+  for (const auto & i : idx) {
+    const auto & cell = grid_ptr_->getCell(i);
+    sum_x += cell.avg_radius_;
+    sum_y += cell.avg_height_;
+    sum_xy += cell.avg_radius_ * cell.avg_height_;
+    sum_x2 += cell.avg_radius_ * cell.avg_radius_;
+  }
+  const float n = static_cast<float>(idx.size());
+  a = (n * sum_xy - sum_x * sum_y) / (n * sum_x2 - sum_x * sum_x);
+  // limit gradient
+  a = std::clamp(a, -param_.global_slope_max_ratio, param_.global_slope_max_ratio);
+  b = (sum_y - a * sum_x) / n;
+}
+
+void GridGroundFilter::initializeGround(pcl::PointIndices & out_no_ground_indices)
+{
+  std::unique_ptr<ScopedTimeTrack> st_ptr;
+  if (time_keeper_) st_ptr = std::make_unique<ScopedTimeTrack>(__func__, *time_keeper_);
+
+  const auto grid_size = grid_ptr_->getGridSize();
+  // loop over grid cells
+  for (size_t idx = 0; idx < grid_size; idx++) {
+    auto & cell = grid_ptr_->getCell(idx);
+    if (cell.is_ground_initialized_) continue;
+    // if the cell is empty, skip
+    if (cell.isEmpty()) continue;
+
+    // check scan root grid
+    if (cell.scan_grid_root_idx_ >= 0) {
+      const Cell & prev_cell = grid_ptr_->getCell(cell.scan_grid_root_idx_);
+      if (prev_cell.is_ground_initialized_) {
+        cell.is_ground_initialized_ = true;
+        continue;
+      }
+    }
+
+    // initialize ground in this cell
+    bool is_ground_found = false;
+    PointsCentroid ground_bin;
+
+    for (const auto & pt : cell.point_list_) {
+      const size_t & pt_idx = pt.index;
+      const float & radius = pt.distance;
+      const float & height = pt.height;
+
+      const float global_slope_threshold = param_.global_slope_max_ratio * radius;
+      if (height >= global_slope_threshold && height > param_.non_ground_height_threshold) {
+        // this point is obstacle
+        out_no_ground_indices.indices.push_back(pt_idx);
+      } else if (
+        abs(height) < global_slope_threshold && abs(height) < param_.non_ground_height_threshold) {
+        // this point is ground
+        ground_bin.addPoint(radius, height, pt_idx);
+        is_ground_found = true;
+      }
```

Comment:
[imo]

Points where `abs(height) >= global_slope_threshold` but `height <= param_.non_ground_height_threshold` are neither classified as ground nor obstacle. ( This would be underground points like ceiling reflection or road depression.)
Just add `else` statements to clearfy its existence.